streams/src/main/java/org/apache/kafka/streams/state/internals/buffer/InMemoryTimeOrderedKeyValueBufferStore.java

```diff
@@ -0,0 +1,219 @@
+/*
+ * Licensed to the Apache Software Foundation (ASF) under one or more
+ * contributor license agreements. See the NOTICE file distributed with
+ * this work for additional information regarding copyright ownership.
+ * The ASF licenses this file to You under the Apache License, Version 2.0
+ * (the "License"); you may not use this file except in compliance with
+ * the License. You may obtain a copy of the License at
+ *
+ *    http://www.apache.org/licenses/LICENSE-2.0
+ *
+ * Unless required by applicable law or agreed to in writing, software
+ * distributed under the License is distributed on an "AS IS" BASIS,
+ * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+ * See the License for the specific language governing permissions and
+ * limitations under the License.
+ */
+package org.apache.kafka.streams.state.internals.buffer;
+
+import org.apache.kafka.clients.consumer.ConsumerRecord;
+import org.apache.kafka.common.header.Header;
+import org.apache.kafka.common.header.Headers;
+import org.apache.kafka.common.header.internals.RecordHeaders;
+import org.apache.kafka.common.serialization.ByteArraySerializer;
+import org.apache.kafka.common.serialization.BytesSerializer;
+import org.apache.kafka.common.utils.Bytes;
+import org.apache.kafka.streams.KeyValue;
+import org.apache.kafka.streams.processor.ProcessorContext;
+import org.apache.kafka.streams.processor.StateStore;
+import org.apache.kafka.streams.processor.internals.ProcessorRecordContext;
+import org.apache.kafka.streams.processor.internals.ProcessorStateManager;
+import org.apache.kafka.streams.processor.internals.RecordBatchingStateRestoreCallback;
+import org.apache.kafka.streams.processor.internals.RecordCollector;
+import org.apache.kafka.streams.state.StoreBuilder;
+import org.apache.kafka.streams.state.internals.ContextualRecord;
+import org.slf4j.Logger;
+import org.slf4j.LoggerFactory;
+
+import java.nio.ByteBuffer;
+import java.util.Collection;
+import java.util.Collections;
+import java.util.LinkedList;
+import java.util.Map;
+
+public class InMemoryTimeOrderedKeyValueBufferStore implements StateStore {
+    private static final Logger LOG = LoggerFactory.getLogger(InMemoryTimeOrderedKeyValueBufferStore.class);
+    private static final String BUFFER_TIME_HEADER = "_bufferTime";
```

Comment:
Is there some convention we use for header names?

Reply:
Not yet, because Streams does not use headers yet. However: So we want to use header? We could encode it in the value directly instead.

If we really want to use headers, should we use something "short" to save storage space? It's 11 bytes overhead per record in the changelog topic using `_bufferTime`.

Reply:
yeah, I was thinking "_bt" for short, but didn't want to lead with something too cryptic.

Looking at the implementation again, though, I think I'll just ditch the header and pack it in the value. I'm remembering now that this was less convenient when the value was generic, but it's really straightforward now.